server/bin/pbench-restore-and-unpack-tarballs.py

```diff
@@ -0,0 +1,243 @@
+            file=sys.stderr,
+        )
+        return 4
+
+    backup = config.conf.get("pbench-server", "pbench-backup-dir")
+    backup_p = Path(backup).resolve(strict=True)
+
+    if not backup_p.is_dir():
+        print(
+            "The configured pbench-backup-dir directory, {backup}, is not a"
+            " valid directory",
+            file=sys.stderr,
+        )
+        return 6
+
+    start = pbench._time()
+
+    if options.dry_run:
+        print(f"Restoring tar balls from backup volume, {backup} (started at: {start})")
+
+    gen = gen_list(backup_p)
+    tbs_sorted = sorted(
+        [
+            (tb_dt, controller_name, tar_ball_name)
+            for tb_dt, controller_name, tar_ball_name in gen
+        ],
+        reverse=True,
+    )
+    tbs_cnt = len(tbs_sorted)
+    tbs_left = tbs_cnt
+    tbs_existing = 0
+    tbs_restored = 0
+    for tb_dt, ctrl, tb in tbs_sorted:
+        try:
+            a_tb = archive_p / ctrl / tb.name
+            tbs_left -= 1
+
+            if a_tb.exists():
+                tbs_existing += 1
+                print(
+                    f"Exists {tbs_existing:d} ({tbs_restored:d} restored) of (tbs_cnt:d): {a_tb}",
+                    flush=True,
+                )
+                continue
+
+            # Create controller directory if it doesn't exist
+            ctrl_p = archive_p / ctrl
+            if not options.dry_run:
+                try:
+                    ctrl_p.mkdir()
+                except FileExistsError:
+                    pass
+                else:
+                    print(f"\nCreated controller directory, '{ctrl_p}'", flush=True)
+
+            # Copy tar ball to archive tree, along with its .md5 ...
+            cp_cmd = f"cp -a {tb} {tb}.md5 {ctrl_p}/"
+            print(f"\n{cp_cmd}", flush=True)
+            if not options.dry_run:
+                cp = subprocess.run(cp_cmd, shell=True, stderr=subprocess.STDOUT)
+                if cp.returncode != 0:
+                    print(
+                        f"FAILURE: cp command: '{cp_cmd}': {cp.returncode}, {cp.stdout!r}",
+                        file=sys.stderr,
+                    )
+                    sys.exit(1)
+            # ... and run md5sum check against it
+            md5sum_cmd = f"md5sum --check {a_tb}.md5"
+            print(md5sum_cmd, flush=True)
+            if not options.dry_run:
+                cp = subprocess.run(md5sum_cmd, shell=True, cwd=str(ctrl_p))
+                if cp.returncode != 0:
+                    print(
+                        f"FAILURE: md5sum command: '{md5sum_cmd}': {cp.returncode}, {cp.stdout!r}",
+                        file=sys.stderr,
+                    )
+                    sys.exit(1)
+
+            # Create the symlink recording the tar ball is already backed up
+            backed_up_dir = ctrl_p / "BACKED-UP"
+            if not options.dry_run:
+                backed_up_dir.mkdir(exist_ok=True)
```

Comment:
No because if we are restoring hundreds of tar balls in one directory we'll see hundreds of directory creations, and that made it a pain to wade through the output to verify it was working.

Reply:
It's not hard for the dry-run branch to keep track of which would-create messages it has printed (you just use a `dict` keyed by the path).

Reply:
Seems like a lot of added mechanism for little gain, since the directory creation is implied by the tar balls that would be restored in the dry-run.